### ISSUE TITLE
GetComponent  - performance w.r.t. reflect

### ISSUE THREAD

**Issue by EtienneBruines** [paked/engi#55]
The `GetComponent` function we use, is quite nice. The current usage being:

```go
var comp *SpaceComponent
if !ent.GetComponent(&comp) {
    return // not found
}
```

However, `reflect` is quite slow. 

An alternative:
```go
var comp *SpaceComponent
interf := ent.GetComponent(comp)
if comp, ok = interf.(*SpaceComponent); !ok {
    return // not found
}
```
But ... This does not feel right, but rewriting it as this isn't perfect either:

```go
var comp *SpaceComponent
var ok bool
comp, ok = ent.GetComponent(comp).(*SpaceComponent)
if !ok {
    return // not found
} 
```

Or this:

```go
var comp *SpaceComponent
if comp, ok = ent.GetComponent(comp).(*SpaceComponent); !ok {
    return // not found
}
```

Ignoring the `reflect` package and try to manipulate it with the `unsafe` package also didn't quite work. 

The fastest solution I've come up so far, is this one: 

```go
// SPACE_COMPONENT is a uint32 or something, thats generated 
// on build-time by the user; the same would be needed for AddComponent, etc.
// or on boot-time, when calling some kind of Register function
comp, ok :=  ent.GetComponent(SPACE_COMPONENT).(*SpaceComponent)
if !ok {
    return // not found
}
```
----------

I'm not sure what we should do here; if we should do anything at all. All I know for now, is that 20% of the runtime, Go is occupied with the `reflect` part of the `GetComponent` function (i.e.: that's about 95% of the `GetComponent` function duration). 

I'm putting it here, because perhaps someone has a nice suggestion as to how to improve this.

**Comment by paked** [comment]
Before we make any assumptions. I'd like to question how slow it actually is, and if it would *actually* effect out games? Can we get a benchmark on this?

**Comment by EtienneBruines** [comment]
What would you like to see benchmarked? I've benchmarked the `CollisionSystem`, and there it was mostly occupied with `GetComponent`.

**Comment by EtienneBruines** [comment]
```
      flat  flat%   sum%        cum   cum%
    1100ms 13.75% 13.75%     1420ms 17.75%  runtime.mallocgc
     800ms 10.00% 23.75%     1030ms 12.88%  github.com/paked/engi.(*Batch).Draw
     590ms  7.38% 31.13%     2650ms 33.12%  github.com/paked/engi.(*Entity).GetComponent
```

Note the `cum%` column; this is the percentage of the total runtime, it's busy doing `GetComponent` (or any function called by `GetComponent`).

**Comment by EtienneBruines** [comment]
I just thought of the best mix between usage and performance. 

```go
var comp *CollisionComponent
if comp, ok = comp.Type(comp.Type()); !ok {
    return // component not found
}
```

Thoughts? @paked @matiwinnetou

**Comment by paked** [comment]
@EtienneBruines Did you mean?

```go
var comp *CollisionComponent
if comp, ok = entity.Type(comp.Type()); !ok {
    return // component not found
}
```

**Comment by EtienneBruines** [comment]
I did; already corrected.

**Comment by paked** [comment]
@EtienneBruines I like that idea, what about:

```go
var comp *CollisionComponent
if  !entity.GetComponent(comp.Type(), comp) {
    return // component not found
}
```

Trying to get rid of reassigning comp...

**Comment by EtienneBruines** [comment]
`GetComponent` can only receive an `interface{}` for each `*Component`. And if you want to assign to an interface, you (unfortunately) have to use `reflect`.

**Comment by paked** [comment]
I'm *really* not a fan of type assertions.

I'd like to go as long as possible without them. Do we have statistics from @matiwinnetou's galaxian yet?

**Comment by EtienneBruines** [comment]
@paked Here they are https://github.com/paked/engi/issues/55#issuecomment-153979326

**Comment by EtienneBruines** [comment]
Okay, the last snippet I put down, doesn't work. Even when fixing all the null-pointer exceptions (for which I still havent found the example), `c` does not change outside the function call. 

This means we do either have to use reflect to assign it to an `interface{}` anyway, or return the Component, and then type-cast it. Not sure why you're not a fan of it, though.

**Comment by EtienneBruines** [comment]
OK, the benchmarks aren't what I expected them to be. The question is: is this 20% worth it?

```
BenchmarkReflect-8      	20000000	       625 ns/op
BenchmarkTypeAssertion-8	30000000	       501 ns/op
```

**Comment by paked** [comment]
On IRC @EtienneBruines proposed a hybrid solution. This may be the best option to go with. As both options require the exact same internal data structures.

Suggestions for this second function were

* `Interface`
* `ComponentFast`

**Comment by EtienneBruines** [comment]
I do like the sound of `ComponentFast`. 

Then that leaves us with: which one should the `engi core` use? I would say the core should be as fast as possible, with the demos showing the usage of the one more readable. 

Thoughts?

**Comment by paked** [comment]
@EtienneBruines SGTM

**Comment by EtienneBruines** [comment]
Fixed as of #73.
